### PR TITLE
Delay missing tx envelopes

### PR DESCRIPTION
**Pull request by omerfirmak**

### CODE REVIEW

**Comment by omerfirmak** [comment]
Added some commits to speed up our envelope handling, I am seeing some great numbers on my local runs.

File: source/agora/consensus/Ledger.d [hunk 1]
```diff
@@ -282,6 +282,21 @@ public class NodeLedger : Ledger
         return this.unknown_txs;
     }
 
+    /***************************************************************************
+
```
Comment (linked0) [inline source/agora/consensus/Ledger.d:285]: Needs function description

File: source/agora/consensus/pool/Transaction.d [hunk 2]
```diff
@@ -537,6 +542,36 @@ public class TransactionPool
         return null;
     }
 
+    /***************************************************************************
+
```
Comment (linked0) [inline source/agora/consensus/pool/Transaction.d:546]: Needs function description.

File: source/agora/consensus/pool/Transaction.d [hunk 3]
```diff
@@ -537,6 +542,36 @@ public class TransactionPool
...
+     ***************************************************************************/
+
+    public Hash[] getUnknownTXsFromSet (Set!Hash hashes) @trusted nothrow
+    {
+        return hashes[].filter!(hash => !this.hasTransactionHash(hash)).array;
+    }
+
+    /***************************************************************************
+
```
Comment (linked0) [inline source/agora/consensus/pool/Transaction.d:561]: Needs function description.

File: source/agora/consensus/pool/Transaction.d [hunk 4]
```diff
@@ -537,6 +542,36 @@ public class TransactionPool
...
+        Params:
+            hashes = set of tx hashes
+
+        Returns:
+            Transactions in the pool that match one of the hashes in the list
```
Comment (linked0) [inline source/agora/consensus/pool/Transaction.d]: Needs correct description

File: source/agora/network/Manager.d [hunk 5]
```diff
@@ -764,6 +764,21 @@ public class NetworkManager
         return this.peers.filter!(p => p.isAuthenticated());
     }
 
+    /***************************************************************************
+
```
Comment (linked0) [inline source/agora/network/Manager.d:768]: Needs function and parameter description

File: source/agora/consensus/protocol/Nominator.d [hunk 6]
```diff
@@ -726,11 +726,66 @@ extern(D):
...
+            return;
+
         if (this.scp.receiveEnvelope(shared_env) != SCP.EnvelopeState.VALID)
             log.trace("SCP indicated invalid envelope: {}", scpPrettify(&envelope, &this.getQSet));
         else
-            this.emitEnvelope(envelope);
+            this.emitEnvelope(shared_env.getEnvelope());
+    }
+
+    ///
```
Comment (linked0) [inline source/agora/consensus/protocol/Nominator.d]: DDOC please

**Review by linked0 — COMMENTED** [review]
Broadly LGTM.
Please add commit details on the commits that don't have any.

File: source/agora/consensus/Ledger.d (hunk 1 above)
Comment (omerfirmak) [reply source/agora/consensus/Ledger.d:285]: I believe the `Returns` doc is enough

File: source/agora/consensus/pool/Transaction.d (hunk 2 above)
Comment (omerfirmak) [reply source/agora/consensus/pool/Transaction.d:546]: Ditto

File: source/agora/consensus/pool/Transaction.d (hunk 3 above)
Comment (omerfirmak) [reply source/agora/consensus/pool/Transaction.d:561]: Ditto

File: source/agora/consensus/pool/Transaction.d (hunk 4 above)
Comment (omerfirmak) [reply source/agora/consensus/pool/Transaction.d]: Updated

File: source/agora/network/Manager.d (hunk 5 above)
Comment (omerfirmak) [reply source/agora/network/Manager.d:768]: Updated

File: source/agora/consensus/protocol/Nominator.d (hunk 6 above)
Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d]: Updated

**Comment by omerfirmak** [comment]
MacOS green

**Review by hewison-chris — APPROVED** [review]
LGTM

**Review by linked0 — APPROVED** [review]
LGTM. There is an error as follows tho, which is showing sometimes these days.

```
Unittesting agora.test.Fee (took 20 secs, 919 ms, 29 μs, and 9 hnsecs)
libc++abi: terminating with uncaught exception of type std::runtime_error: moved to a bad state (ballot protocol)
./ci/run.sh: line 21:  3202 Abort trap: 6           ./build/agora-unittests
```

**Pull request merged**